homeassistant/components/ecovacs/sensor.py

```diff
@@ -0,0 +1,269 @@
+        translation_key="stats_total_time",
+        native_unit_of_measurement=UnitOfTime.HOURS,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    EcovacsSensorEntityDescription[TotalStatsEvent](
+        capability_fn=lambda caps: caps.stats.total,
+        value_fn=lambda e: e.cleanings,
+        key="stats_total_cleanings",
+        translation_key="stats_total_cleanings",
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    EcovacsSensorEntityDescription[BatteryEvent](
+        capability_fn=lambda caps: caps.battery,
+        value_fn=lambda e: e.value,
+        key=ATTR_BATTERY_LEVEL,
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    EcovacsSensorEntityDescription[NetworkInfoEvent](
+        capability_fn=lambda caps: caps.network,
+        value_fn=lambda e: e.ip,
+        key="wifi_ip",
+        translation_key="wifi_ip",
+        entity_registry_enabled_default=False,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    EcovacsSensorEntityDescription[NetworkInfoEvent](
+        capability_fn=lambda caps: caps.network,
+        value_fn=lambda e: e.rssi,
+        key="wifi_rssi",
+        translation_key="wifi_rssi",
+        entity_registry_enabled_default=False,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    EcovacsSensorEntityDescription[NetworkInfoEvent](
+        capability_fn=lambda caps: caps.network,
+        value_fn=lambda e: e.ssid,
+        key="wifi_ssid",
+        translation_key="wifi_ssid",
+        entity_registry_enabled_default=False,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+)
+
+
+_SUPPORTED_LIFE_SPAN_TYPES = (
+    LifeSpan.BRUSH,
+    LifeSpan.FILTER,
+    LifeSpan.SIDE_BRUSH,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Add entities for passed config_entry in HA."""
+    controller: EcovacsController = hass.data[DOMAIN][config_entry.entry_id]
+
+    entities: list[EcovacsEntity] = get_supported_entitites(
+        controller, EcovacsSensor, ENTITY_DESCRIPTIONS
+    )
+    for device in controller.devices:
+        lifespan_capability = device.capabilities.life_span
+        for component in _SUPPORTED_LIFE_SPAN_TYPES:
+            if component in lifespan_capability.types:
+                entities.append(
+                    EcovacsLifeSpanSensor(device, lifespan_capability, component)
+                )
+
+        if capability := device.capabilities.error:
+            entities.append(EcovacsLastErrorSensor(device, capability))
+
+    async_add_entities(entities)
+
+
+class EcovacsSensor(
+    EcovacsDescriptionEntity[CapabilityEvent, EcovacsSensorEntityDescription],
+    SensorEntity,
+):
+    """Ecovacs sensor."""
+
+    entity_description: EcovacsSensorEntityDescription
```

Comment:
@cdce8p I hope it is fine to ping you directly :)

I have a type issue with mypy.

Without this code line I get the following error:
`homeassistant/components/ecovacs/sensor.py:176: error: Definition of "entity_description" in base class "EcovacsDescriptionEntity" is incompatible with definition in base class "SensorEntity"  [misc]`

But I use generic in the base class:
```python
class EcovacsDescriptionEntity(
    EcovacsEntity[CapabilityT], Generic[CapabilityT, _EntityDescriptionT]
):
    """Ecovacs entity."""

    entity_description: _EntityDescriptionT
```

What do I miss?

Reply:
Try removing the `entity_description: _EntityDescriptionT` line from `EcovacsDescriptionEntity`.

The issue here is that you're trying to make a variable generic which isn't generic in the base class. `entity_description: ...` is basically a just hack to avoid having to make it generic in the first place and needing to update all integrations. It would be "more" correct and at some point we might even want to do that. I would wait for [PEP 696](https://peps.python.org/pep-0696/) support though (at least).

As for your case, I would suggest to remove `_EntityDescriptionT` and continue to use the "hack" here. That should probably work.

Reply:
Thanks :)